champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/knight/passives/Thorns.java

```diff
@@ -67,31 +73,37 @@ public SkillType getType() {
         return SkillType.PASSIVE_A;
     }
 
-    @EventHandler
+    @EventHandler(priority = EventPriority.HIGHEST)
     public void onDamage(CustomDamageEvent event) {
         if (event.isCancelled()) return;
         if (event.getCause() != DamageCause.ENTITY_ATTACK) return;
-        if (!(event.getDamagee() instanceof Player p)) return;
+        if (!(event.getDamagee() instanceof Player player)) return;
         if (event.getDamager() == null) return;
 
-        int level = getLevel(p);
-        if (level > 0) {
-            LivingEntity damager = event.getDamager();
-            if (!cd.containsKey(damager)) {
-                cd.put(damager, System.currentTimeMillis());
-            } else {
-                if(UtilTime.elapsed(cd.get(damager), (long) (internalCooldown * 1000L))){
-                    UtilDamage.doCustomDamage(new CustomDamageEvent(damager, p, null, DamageCause.CUSTOM, getDamage(level), false, getName()));
-                    cd.put(damager, System.currentTimeMillis());
-                }
-            }
+        if(getLevel(player) < 1) return;
+
+        LivingEntity damager = event.getDamager();
+        hitTimestamps.putIfAbsent(player, new LinkedList<>());
+        Queue<Long> timestamps = hitTimestamps.get(player);
+
+        timestamps.offer(System.currentTimeMillis());
+
+        while (!timestamps.isEmpty() && System.currentTimeMillis() - timestamps.peek() > (reflectTime * 1000)) {
+            timestamps.poll();
+        }
```

Comment:
I would rather you just do this in an update event and process 1 per player per tick or 2 ticks.

Since LivingEntity is also nullable, migrate to a weakhashmap

Reply:
Just curious, whats problematic about this approach?

Reply:
While you have done it in a safe way, while loops can cause the server to hang if they get stuck iterating. Since minecraft runs mainly on a single thread, I think its a safer bet to avoid them all together

Reply:
This doesn't just apply to while loops, but any kind of loop

Reply:
> This doesn't just apply to while loops, but any kind of loop. Typically we loop over a fixed size or an existing list, but in this case you would be relying on the timestamp to be correctly removed before it exits

in that case there are definitely a few while loops in other classes that should be updated in this way.